sql,cli: improve stmt bundles for MR tables

yuzefovich opened the pull request:
This PR contains a few commits that improve the bundle collection and the bundle recreation for MR tables. See each commit for details.

Fixes: #99131.
Fixes: #123897.
Fixes: #138015.

yuzefovich committed:
sql: remove redundant queryRows helper

This was used only when querying for CREATE TYPE stmt when building the
stmt bundle, and there we expect exactly one row with a single string
value. There is an exception though that `crdb_internal_region` type
won't show up in the vtable, so we need to allow for empty result to be
returned.

This commit also fixes a minor bug that we forgot to add a semicolon to
CREATE TYPE statements.

Release note: None

yuzefovich committed:
cli: improve `debug sb recreate` for multi-region bundles

This commit adjusts the stmt bundle recreation logic to start multiple
nodes with specified localities in some cases. In particular, after the
previous commit we now include CREATE DATABASE stmts into `schema.sql`
file with the necessary region information, so this commit parses the
schema file and extracts all regions found among all CREATE db stmts.
Regions are deduplicated, and then we start a single node for each
region we found.

Release note: None

yuzefovich committed:
cli: set the current database on `debug sb recreate`

This commit adjusts the bundle recreate command so that the correct
`database` session variable is set. This will make it easier to run the
query. It required a minor adjustment to no longer special case this
session variable when creating the bundle - we simply execute it after
creating all schema objects on recreation.

Release note: None

yuzefovich committed:
revert "sql: include DB name in `create_statement` in `create_type_statements`"

This reverts commit 644c458333e05bbb821dab1b9bb693d8003ef4fa.

The main rationale for the revert is that this affects SHOW CREATE ALL
TYPES making the result less portable. The following commit will modify
the bundle construction in a different way so that fully-qualified UDT
create statement is not needed.

yuzefovich committed:
sql: improve stmt bundle creation

This commit refactors how we create the stmt bundle to make it work with
cross-database references. Previously, we didn't handle the DB names
properly in all CREATE statements (for example, we adjusted the CREATE
TABLE stmt to use "db.schema.name" notation, but we didn't do it for
CREATE VIEW nor for CREATE SEQUENCE; relatedly, the UDTs in the table
schema were not db-qualified). This commit takes a different approach:
given that all `crdb_internal` virtual tables that we use for the bundle
creation use "schema.name" notation, we now preface every CREATE stmt
with a `USE <dbName>;` stmt. This change is applied to tables,
sequences, views, UDTs, and routines. Databases and schemas already
had the necessary parts, so they are unaffected.

Additionally, this commit adjusts the `CREATE DATABASE` statements to use
`create_statement` column of `crdb_internal.databases` virtual table. The
main benefit is that now multi-region databases will get their regions right
away.

It also picks up minor fixes from
644c458333e05bbb821dab1b9bb693d8003ef4fa to properly use `""` search
path for UDTs and routines as well as adding a cast to OID type.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138023)
<!-- Reviewable:end -->

michae2 reviewed (commented):
Thank you for doing this. It's going to be really helpful! Just some nits.

> Reviewed 6 of 6 files at r1, 1 of 1 files at r2, 5 of 5 files at r3, 4 of 4 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-:-OF8qEax0cWzAVpaHpz2:b-1fe4lh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8p8nW6pxBye9xlYDP:-OF8p8nW6pxBye9xlYDQ:b-o0hjuj):*
This is SO NICE! Thank you for doing this!
___
*[`-- commits` line 71 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8pgIm2pxOTKgEuHHr:-OF8pgIm2pxOTKgEuHHs:b-83sjoh):*
Nice!
___
*[`pkg/cli/statement_bundle.go` line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8lrj580pynd12R5HV:-OF8lrj580pynd12R5HW:by4d9g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac354f04c8bcae2a48dc742a025500b9250a7974/pkg/cli/statement_bundle.go#L130)):*
> ```Go
> 		for _, line := range strings.Split(schema, "\n") {
> 			line = strings.TrimSpace(line)
> 			if strings.HasPrefix(line, "CREATE DATABASE") &&
> ```

Would it be possible to use `parser.ParseOne` here to get a `tree.CreateDatabase` AST node?
___
*[`pkg/cli/interactive_tests/test_sb_recreate_mr.tcl` line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8pV44B2ScBavmcmpW:-OF8pV44B2ScBavmcmpX:b9aws6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac354f04c8bcae2a48dc742a025500b9250a7974/pkg/cli/interactive_tests/test_sb_recreate_mr.tcl#L47)):*
> ```Tcl
> 
> spawn $argv debug sb recreate bundle
> eexpect "Started 6 nodes with regions"
> ```

Really cool! Thank you for writing a test!
___
*[`pkg/sql/explain_bundle.go` line 605 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8gXE70W0eLA2ndv_6:-OF8gXE70W0eLA2ndv_7:b-gjyi13) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L605)):*
> ```Go
> 	// routines.
> 	// TODO(138022): we need to collect DBs and schemas for transitive
> 	// dependencies.
> ```

Thank you for opening these!
___
*[`pkg/sql/explain_bundle.go` line 974 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8_8-5AwndEU2-mQ_o:-OF8_8-5AwndEU2-mQ_p:b-co2bou) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L974)):*
> ```Go
> 
> func getDBName(tn *tree.TableName) (string, error) {
> 	parts := strings.Split(tn.FQString(), ".")
> ```

Will this work if the object name contains periods? Could we call `tn.Catalog()` instead?
___
*[`pkg/sql/explain_bundle.go` line 984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8g9FSA8hMpXvvQOD9:-OF8g9FSA8hMpXvvQODA:buatjel) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L984)):*
> ```Go
> 	// TODO(138015): this might require an adjustment when the DB name has
> 	// special characters.
> 	fmt.Fprintf(w, "USE %s;\n", dbName)
> ```

Could we use `tree.NameString()` or something to properly escape here?
___
*[`pkg/sql/explain_bundle.go` line 1099 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8aoR3DMUV2v64kyfR:-OF8aoR3DMUV2v64kyfS:bxab6hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L1099)):*
> ```Go
> 		}
> 		createStatement, err := c.query(fmt.Sprintf(
> 			"SELECT create_statement FROM crdb_internal.databases WHERE name = '%s'", db,
> ```

Will this work if the database name contains a single-quote character? This looks like a SQL injection vector. Could we use `tree.NameString()` to properly escape?
___
*[`pkg/sql/explain_bundle.go` line 1115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8hl0_7KctwILhcUfn:-OF8hl0_7KctwILhcUfo:b-s6evtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L1115)):*
> ```Go
> 		// TODO(138015): this logic assumes that we don't have the period in the
> 		// database name.
> 		_, schemaOnly, found := strings.Cut(schema, ".")
> ```

Maybe `schemaNames` should be `map[string]tree.Name` or something?
___
*[`pkg/cli/interactive_tests/test_sb_recreate.tcl` line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8iH1z3Rkmlf1iZmCt:-OF8iH1z3Rkmlf1iZmCu:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/cli/interactive_tests/test_sb_recreate.tcl#L53)):*
> ```Tcl
> system "$python $pyfile stmt-bundle-$id1.zip bundle"
> 
> spawn $argv debug sb recreate bundle
> ```

Nice test!
___
*[`pkg/cli/statement_bundle.go` line 197 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8ptde5T1NEpx5fnoX:-OF8ptde5T1NEpx5fnoY:by7czy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e41a7490960bae34f972f2550c23d34e29302e17/pkg/cli/statement_bundle.go#L197)):*
> ```Go
> 		for i, stmt := range setStmts {
> 			stmt = strings.TrimSpace(stmt)
> 			if strings.HasPrefix(stmt, "SET database") {
> ```

nit: It might be worth sticking a space at the end just in case we ever create a session var starting with "database_"...


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: handle DB and schema names with special characters in bundles

This commit makes it so that we properly handle database and schema
names that have special characters like periods and single quotes (that
require double-quoting in some contexts). It is achieved by using `tree.Name`
type instead of `string` and relying on the `Format` implementation to
properly escape identifiers or explicitly calling `EscapeSQLString`,
depending on the context.

Release note: None

yuzefovich reviewed (commented):
Thanks for the careful review and suggestions! I added another commit that should handle the special characters in the DB and schema names. RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-:-OFxh57A6v9gUsLuN_TV:b-o521th)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/cli/statement_bundle.go` line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8lrj580pynd12R5HV:-OFxAw-n60Fhp6QrNr60:b6rrz3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac354f04c8bcae2a48dc742a025500b9250a7974/pkg/cli/statement_bundle.go#L130)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would it be possible to use `parser.ParseOne` here to get a `tree.CreateDatabase` AST node?
</blockquote></details>

Nice, I didn't think of it, done.
___
*[`pkg/cli/statement_bundle.go` line 197 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8ptde5T1NEpx5fnoX:-OFx6QMj9Rc0aY92a1up:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e41a7490960bae34f972f2550c23d34e29302e17/pkg/cli/statement_bundle.go#L197)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: It might be worth sticking a space at the end just in case we ever create a session var starting with "database_"...
</blockquote></details>

Good point, done.
___
*[`pkg/sql/explain_bundle.go` line 974 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8_8-5AwndEU2-mQ_o:-OFxeGTe3_u9yzdUqvXA:b-7l7s90) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L974)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Will this work if the object name contains periods? Could we call `tn.Catalog()` instead?
</blockquote></details>

No, and since this was already broken elsewhere I initially accepted this, but now it should work.
___
*[`pkg/sql/explain_bundle.go` line 984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8g9FSA8hMpXvvQOD9:-OFxePuf9k8VaAPhtuvN:b-vbxs2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L984)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Could we use `tree.NameString()` or something to properly escape here?
</blockquote></details>

Thanks, switched to using `tree.Name`.
___
*[`pkg/sql/explain_bundle.go` line 1099 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8aoR3DMUV2v64kyfR:-OFxdJJi5kDoiQPYOzYs:b-k48tsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L1099)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Will this work if the database name contains a single-quote character? This looks like a SQL injection vector. Could we use `tree.NameString()` to properly escape?
</blockquote></details>

Good point, this would have been busted. I refactored the code to manually escape single quote characters - looks like in `lexbase` we don't handle single quotes in a special way because we'd surround the whole string value with double quotes, but in this case we need a string literal surrounded by single quotes. Let me know if you can think of a cleaner way to do this.
___
*[`pkg/sql/explain_bundle.go` line 1115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8hl0_7KctwILhcUfn:-OFxdI8iEEhlXQ2CzAPg:bo3vl6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L1115)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Maybe `schemaNames` should be `map[string]tree.Name` or something?
</blockquote></details>

Thanks, done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This looks great!

> Reviewed 9 of 9 files at r6, 1 of 1 files at r7, 5 of 5 files at r8, 4 of 4 files at r9, 5 of 5 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-:-OG29O__77cff4OGxGen:bypmzeb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/explain_bundle.go` line 1099 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8aoR3DMUV2v64kyfR:-OG26nGQC3kOxZHWiGic:b-jtj6l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L1099)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Good point, this would have been busted. I refactored the code to manually escape single quote characters - looks like in `lexbase` we don't handle single quotes in a special way because we'd surround the whole string value with double quotes, but in this case we need a string literal surrounded by single quotes. Let me know if you can think of a cleaner way to do this.
</blockquote></details>

It looks like `tree.(*Name).Format` is calling `lexbase.EncodeRestrictedSQLIdent` on the underlying string, which is where the double quotes come from. I think we want to call `lexbase.EscapeSQLString` on the underlying string instead? So maybe `dbName := lexbase.EscapeSQLString(string(db))` would work?
___
*[`pkg/sql/explain_bundle.go` line 898 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OG260Ab6Fj09vEvcQUC:-OG260Ab6Fj09vEvcQUD:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/245206caff597b38b535a0bc31fe5a6a399e6bf7/pkg/sql/explain_bundle.go#L898)):*
> ```Go
> 		}
> 		if _, ok := sessionVarNeedsEscaping[varName]; ok {
> 			value = lexbase.EscapeSQLString(value)
> ```

Nice!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the careful review!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-:-OG2bNMAAOtV3VEs7HX9:b-ae1x1u)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`pkg/sql/explain_bundle.go` line 1099 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/138023#-OF8aoR3DMUV2v64kyfR:-OG2bKUb7svQmy_YZC9_:b-1jib13) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd7abfc1200d755a9edbf0fe99655540028e1e9/pkg/sql/explain_bundle.go#L1099)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It looks like `tree.(*Name).Format` is calling `lexbase.EncodeRestrictedSQLIdent` on the underlying string, which is where the double quotes come from. I think we want to call `lexbase.EscapeSQLString` on the underlying string instead? So maybe `dbName := lexbase.EscapeSQLString(string(db))` would work?
</blockquote></details>

Indeed, nice, thanks!


<!-- Sent from Reviewable.io -->
